book/src/proptest-derive/errors.md

```diff
@@ -0,0 +1,631 @@
+
+A trivial example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Uninhabited {
+    a: u32,
+    b: !,
+}
+```
+
+Because there exist no values assignable to field `b`, it is also impossible to
+construct an instance of struct `Uninhabited`.
+
+Proptest's ability to identify uninhabited types is limited. If it does not
+recognise a particular type, it will attempt to use it like any other type and
+you will instead get an error about the type not implementing `Arbitrary`
+trait.
+
+## E0004
+
+This error occurs when `#[derive(Arbitrary)]` is used on an enum with no
+variants at all. For example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum Uninhabited { }
+```
+
+Such an enum has no values at all, so it does not make sense to provide an
+`Arbitrary` implementation for it.
+
+## E0005
+
+This error occurs if `#[derive(Arbitrary)]` is used on an enum whose variants
+are all uninhabited, using the same logic as described for [`E0003`](#e0003).
+As a result, the enum itself is totally uninhabited.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum Uninhabited {
+    V1(!),
+    V2(!, !),
+}
+```
+
+## E0006
+
+This error occurs if `#[derive(Arbitrary)]` is used on an enum where all
+inhabited variants are marked with `#[proptest(skip)]`. In other words,
+proptest is forbidden from generating any of the enum's variants, and thus the
+enum itself is unconstructable.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+enum MyEnum {
+    // Ordinarily, proptest would be able to generate either of these variants,
+    // but both are forbidden, so in the end proptest isn't allowed to generate
+    // anything at all.
+    #[proptest(skip)]
+    V1,
+    #[proptest(skip)]
+    V2(u32),
+    // This variant is implicitly skipped because proptest knows it is
+    // uninhabited.
+    V3(!),
+}
+```
+
+## E0007
+
+This error happens if an attribute `#[proptest(strategy = "expr")]` or
+`#[proptest(value = "expr")]` is applied to an item such as an enum or struct.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+#[proptest(value = "MyStruct(42)")]
+struct MyStruct(u32);
+```
+
+These attributes are only applicable to fields.
+
+## E0008
+
+This error happens if `#[proptest(skip)]` is applied to an unskippable item.
+For example, struct fields cannot be skipped because Rust requires every field
+of a struct to have a value.
+
+Example:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct WidgetContainer {
+    desired_widget_count: usize,
+    #[proptest(skip)]
+    widgets: Vec<Widget>,
+}
+```
+
+In general, the appropriate way to request proptest to not generate a field
+value is to use `#[proptest(value = "expr")]` to provide a fixed value
+yourself. For example, the above code could be properly written as follows:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct WidgetContainer {
+    desired_widget_count: usize,
+    #[proptest(value = "vec![]")] // Always generate an empty widget vec
+    widgets: Vec<Widget>,
+}
+```
+
```

Comment:
```suggestion

The only things that can be skipped are enum variants. For example:

```rust
#[derive(Debug, Arbitrary)]
enum Foo {
    NormalVariant(u8),
    #[proptest(skip)]
    MySpecialVariant(String),
}
```

You'll need to add ``` to this because GitHub won't let me add them...

Reply:
This seems a bit extraneous; it's not relevant to the user's problem because they're clearly not working with an enum, and switching to using an enum is not a sensible solution.

Reply:
Hmm; I suppose -- but showing what the attribute is used for is nice, so maybe link to other parts of the book (once written) that shows the use of the attribute?

Reply:
Yeah, that sounds like a good plan to me.